src/components/common/dialog.tsx

```diff
@@ -0,0 +1,1382 @@
+    }
+
+    function validate() {
+      validate_Person(dlg.value("person"));
+    }
+
+    const dlg = useDialogState(init, validate);
+
+    return (
+      ...
+      <Person value={dlg.value("person")} />
+      ...
+    );
+
+  Here is a pattern for handling types that include alternatives, such
+  as "PersonValue | string".  This could be used to encode either the
+  state for a working Person component, or an excuse message that
+  explains why it can't work.
+
+    function init_Person(...): PersonValue | string {
+      if (some condition)
+        return _("Something wrong.");
+
+      return { ... };
+    }
+
+    function validate_Person(value: DialogValue<PersonValue | string>) {
+      const val = value.get();
+      if (typeof val == "string")
+          return;
+
+      const p_value = value.at(val);
+
+      p_value.sub("first").validate(v => {
+        if (!v)
+          return "First name can't be empty";
+      })
+    }
+
+    export const Person = ({ value } : { value: DialogValue<PersonValue | string> }) => {
+      const val = value.get();
+      if (typeof val == "string")
+          return null;
+
+      const p_value = value.at(val);
+
+      const { country, first, last } = p_value.get();
+      ...
+        <DialogTextInput value={p_value.sub("first")} />
+      ...
+    }
+
+  Note the use of the "value.at()" function to get a handle for a
+  PersonValue that can be used to access the "first" sub value, etc.
+
+ */
+
+import React, { useState } from "react";
+import { useObject, useInit, useOn } from 'hooks';
+import { EventEmitter } from 'cockpit/event';
+
+import cockpit from "cockpit";
+
+import { Button, type ButtonProps } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import { FormGroup, type FormGroupProps, FormHelperText } from "@patternfly/react-core/dist/esm/components/Form";
+import { TextInput, type TextInputProps } from "@patternfly/react-core/dist/esm/components/TextInput";
+import { Alert } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import {
+    HelperText, HelperTextItem, type HelperTextItemProps
+} from "@patternfly/react-core/dist/esm/components/HelperText";
+import { Checkbox } from "@patternfly/react-core/dist/esm/components/Checkbox";
+import {
+    FormSelect, FormSelectOption, type FormSelectProps,
+} from "@patternfly/react-core/dist/esm/components/FormSelect";
+import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
+
+const _ = cockpit.gettext;
+
+type ArrayElement<ArrayType> =
+  ArrayType extends readonly (infer ElementType)[] ? ElementType : never;
+
+interface DialogValidationState {
+    cached_value: unknown;
+    cached_result: string | undefined;
+    timeout_id: number;
+    promise: Promise<void> | undefined;
+    round_id: unknown;
+}
+
+function toSpliced<T>(arr: T[], start: number, deleteCount: number, ...rest: T[]): T[] {
+    const copy = [...arr];
+    copy.splice(start, deleteCount, ...rest);
+    return copy;
```

Comment:
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)